pallets/runtime/src/bridge.rs

```diff
@@ -0,0 +1,303 @@
+            ensure!(<identity::Module<T>>::has_valid_kyc(&did), "recipient DID has no valid KYC");
+            let mut new_pending_txs: BTreeMap<_, Vec<BridgeTx<T::AccountId>>> = BTreeMap::new();
+            for bridge_tx in Self::pending_txs(&did) {
+                match Self::issue(bridge_tx.clone()) {
+                    Ok(None) => Self::deposit_event(RawEvent::Bridged(bridge_tx)),
+                    Ok(Some(PendingTx {
+                        did: to_did,
+                        bridge_tx,
+                    })) => {
+                        let entry = new_pending_txs
+                            .entry(to_did)
+                            .or_default();
+                        entry.push(bridge_tx.clone());
+                        Self::deposit_event(RawEvent::Pending(PendingTx {
+                            did: to_did,
+                            bridge_tx
+                        }));
+                    }
+                    Err(_) => Self::deposit_event(RawEvent::Failed(bridge_tx)),
+                }
+            }
+            for (to_did, txs) in new_pending_txs {
+                if to_did == did {
+                    <PendingTxs<T>>::insert(did, txs);
+                } else {
+                    <PendingTxs<T>>::mutate(to_did, |pending_txs| pending_txs.extend(txs));
+                }
+            }
+            Ok(())
+        }
+
+        /// Handles an approved validator set change transaction proposal.
+        ///
+        /// NOTE: Extrinsics without `pub` are exported too. This function is declared as `pub` only
+        /// to test that it cannot be called from a wrong `origin`.
+        pub fn handle_change_validators(origin, new_validators: T::AccountId) -> DispatchResult {
+            let sender = ensure_signed(origin.clone())?;
+            ensure!(sender == Self::validators(), "should be called by the validator set account");
+            // Update the validator set.
+            <Validators<T>>::put(new_validators.clone());
+            // Remove the record of the ongoing proposal.
+            <ChangeValidatorsProposals<T>>::remove(&new_validators);
+            Self::deposit_event(RawEvent::ValidatorsChanged(new_validators));
+            Ok(())
+        }
+
+        /// Handles an approved bridge transaction proposal.
+        ///
+        /// NOTE: Extrinsics without `pub` are exported too. This function is declared as `pub` only
+        /// to test that it cannot be called from a wrong `origin`.
+        pub fn handle_bridge_tx(origin, bridge_tx: BridgeTx<T::AccountId>) -> DispatchResult {
+            let sender = ensure_signed(origin.clone())?;
+            ensure!(sender == Self::validators(), "should be called by the validator set account");
+            if let Some(PendingTx {
+                did,
+                bridge_tx,
+            }) = Self::issue(bridge_tx.clone())? {
+                <PendingTxs<T>>::mutate(did, |txs| txs.push(bridge_tx.clone()));
```

Comment:
We need a recovery process for the pending tx.

Reply:
Can you please clarify that. Do you mean that `Self::issue` resulting in an `Err` should lead to recovery from balance overflow rather than to a `Failed` event?

Reply:
If a tx goes into pending state for whatever reason, I don't see any way to "resume" it. This will mean that the tokens locked on ethereum that correspond to this pending issuance will forever be stuck.

Reply:
`finalize_pending` can be called by anybody with the KYC-ed identity as the argument. This function can be called regularly or as a hook after the whatever method performs a successful KYC validation.

Reply:
Ah, I completely missed the `finalize_pending` function. My bad.